enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,358 @@
+
+Which resources ultimately get installed for a given cluster would be the set of resources encompassed
+by the CLUSTER_PROFILE(if any), minus any resources explicitly excluded by the excluded capabilities configuration.
+
+Examples of candidate components to be treated as capabilities:
+
+* console
+* imageregistry
+* samples
+* cluster baremetal operator
+* olm/marketplace
+* kube-storage-version-migrator
+* csi-*
+* insights
+* monitoring
+* ???
+
+To distinguish between the user intent of "Give me the default components that are suitable for my install" and
+"do not exclude any components", users can specify `none` as the capability to exclude.  In the event that we
+want the installer to apply some default exclusions in specific scenarios when the user has not expressed any
+specific desire to exclude particular components, this `none` specification will be a way for the user to opt out
+of having those defaults applied.
+
+Alternative:
+If we think it's better to invert it, such that "if you tell us nothing, you get all the components", we can instead
+define a `default` component keyword, such that if you don't specify any exclusions, the installer leaves it alone
+and you get everything, but if you specify `default`, then the installer knows to replace the exclusion list with
+whatever the installer thinks is best for your particular platform/whatever.
+
+
+3) Pass the list of filtered annotations to the CVO.  This is distinct from overrides because overrides
+put the cluster in an unsupported state.  Filtered annotations are supported for upgrades.  The filtered
+components will be listed in the [ClusterVersion](https://github.com/openshift/api/blob/4436dc8be01e8dcd8b250e1b32bb0fbd64ba78ac/config/v1/types_cluster_version.go#L35) object:
+
+```yaml
+  spec:
+    excludedCapabilities:
+    - console
+    - samples
+```
+
+The CVO will filter out(not apply/reconcile) resources that are annotated with `capability.openshift.io/$exclusions`
+
+If a resource has multiple `capability.openshift.io` annotations, then the resource will only be filtered if all
+the annotations on the resource are matched by a configured filter.
+
+If the install-config specified `none` as the excluded component, the CVO list will be empty.  If the
+install-config specified `default`, the CVO list will be whatever the installer chooses to disable for
+the install.  The CVO does not need to be aware of those special keywords.
+
+4) Admin can remove an item from the filtered annotations list, but they cannot add an item to it.  If an
+item is removed, the CVO will apply the previously filtered(skipped) resources to the cluster on the next reconciliation.
+Adding an item to the filtered list is not supported because it requires the component be removed from the
+running cluster which has more significant implications for how all traces of the component are removed.
+
+The currently configured filter list for the CVO should be recorded in telemeter so we can understand
```

Comment:
assuming monitoring is itself not filtered…

Reply:
Is a service monitor CRD always present?

Reply:
> Is a service monitor CRD always present?

with this feature, the service monitor CRD could be annotated as part of the monitoring capability, such that it only got created if the monitoring capability wasn't disabled.  Or it could not be annotated, in which case it would always be created.  

If we did the former then we'd also need to deal w/ all the other service monitor resources that are going to be created (also annotate them so they get filtered when the monitoring capability is disabled).

if we did the latter, we'd have a bunch of service monitor CRs that weren't actually doing anything and might confuse people who expected them to mean something.

so i'd certainly lean towards doing the former. (filter both the CRD and all the CRs, using the monitoring capability keyword)

Reply:
Maybe we allow multiple annotations and if any is provided the resource is disabled? That way the service monitor could have both "monitoring" and "cluster-foo-operator" and if either is in the exclude list the service monitor wouldn't be created.

Reply:
> Maybe we allow multiple annotations and if any is provided the resource is disabled? 

Unfortunately that conflicts with the behavior this EP already proposed for multiple annotations which is "if a resource specifies multiple annotations, you must exclude all of them to exclude the component" 

and that behavior makes more sense from the perspective of the resource is saying "i'm a part of capability A and capability B", so unless you exclude both of those capabilities, you should get the resource.

but you're right, for a resource like a servicemonitor, it would make more sense to say "you only get this resource if both Monitoring *and* $Component are enabled"

need to think about it some more.  This seems like a more common use case, so maybe inverting the behavior to say "excluding any one of the annotations excludes the resource" is more useful.